File: ruby/lib/muxpad/discovery.rb
# frozen_string_literal: true

require "json"
require "pathname"
require "shellwords"
require "yaml"

module Muxpad
  class Discovery
    LIFECYCLE_SCRIPTS = %w[
      dependencies install postinstall postpack postpublish postversion
      preinstall prepack prepare prepublish prepublishOnly preversion version
    ].freeze

    def scripts(root, exclude: [])
      root = File.expand_path(root)
      root_package = read_package(File.join(root, "package.json"))
      return {} unless root_package

      manager = package_manager(root, root_package)
      packages(root, root_package).to_h do |directory, package|
        package_name = package.fetch("name", File.basename(directory)).to_s
        relative = Pathname.new(directory).relative_path_from(Pathname.new(root)).to_s
        prefix = relative == "." ? nil : package_name
        definitions = filtered_scripts(package.fetch("scripts", {})).filter_map do |name, body|
          id = prefix ? "#{prefix}:#{name}" : name
          next if exclude.any? { |pattern| File.fnmatch?(pattern, id, File::FNM_EXTGLOB) }
          command = script_command(manager, name)
          definition = Definition.new(
            id:, name: id, description: body.to_s, command:, directory: relative,
            placement: "window", exit_mode: "keep-on-error", enabled: true, executable: manager
          )
          [id, definition]
        end
        [directory, definitions]
      end.values.flatten(1).to_h
    rescue Errno::ENOENT, Errno::EACCES, JSON::ParserError, Psych::Exception
      {}
    end

    private

    def packages(root, root_package)
      entries = [[root, root_package]]
      workspace_patterns(root, root_package).each do |pattern|
        next if pattern.start_with?("!")
        Dir.glob(File.join(root, pattern, "package.json")).sort.each do |path|
          next unless Pathname.new(path).expand_path.to_s.start_with?(root + File::SEPARATOR)
          package = read_package(path)
          entries << [File.dirname(path), package] if package
        end
      end
      entries.uniq { |directory, _| directory }
    end

    def workspace_patterns(root, root_package)
      workspaces = root_package["workspaces"]
      patterns = case workspaces
      when Array then workspaces
      when Hash then Array(workspaces["packages"])
      else []
      end
      pnpm_path = File.join(root, "pnpm-workspace.yaml")
      if File.exist?(pnpm_path)
        begin
          pnpm = YAML.safe_load_file(pnpm_path, permitted_classes: [], aliases: false) || {}
          patterns += Array(pnpm["packages"])
        rescue Psych::Exception
          # A broken workspace file should not hide valid root-package scripts.
        end
      end
      patterns.map(&:to_s).uniq
    end

    def filtered_scripts(value)
      return {} unless value.is_a?(Hash)
      names = value.keys.map(&:to_s)
      value.to_h.reject do |name, _|
        name = name.to_s
        LIFECYCLE_SCRIPTS.include?(name) || hook_for_existing_script?(name, names)
      end
    end

    def hook_for_existing_script?(name, names)
      %w[pre post].any? { |prefix| name.start_with?(prefix) && names.include?(name.delete_prefix(prefix)) }
    end

    def package_manager(root, package)
      declared = package["packageManager"].to_s
      declared_manager = declared.split("@", 2).first
      return declared_manager if %w[pnpm yarn bun npm].include?(declared_manager)
      return "pnpm" if File.exist?(File.join(root, "pnpm-lock.yaml"))
      return "yarn" if File.exist?(File.join(root, "yarn.lock"))
      return "bun" if %w[bun.lock bun.lockb].any? { |file| File.exist?(File.join(root, file)) }
      "npm"
    end

    def script_command(manager, name)
      name = Shellwords.escape(name.to_s)
      manager == "npm" ? "npm run #{name}" : "#{manager} #{name}"
    end

    def read_package(path)
      value = JSON.parse(File.read(path))
      value if value.is_a?(Hash)
    rescue Errno::ENOENT, Errno::EACCES, JSON::ParserError
      nil
    end
  end
end

File: internal/discovery/discovery.go
package discovery

import (
	"bytes"
	"encoding/json"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"

	"github.com/Martin-Nyaga/muxpad/internal/config"
	"github.com/Martin-Nyaga/muxpad/internal/shellwords"
	"gopkg.in/yaml.v3"
)

type Discovery struct{}

type script struct {
	Name string
	Body string
}

type packageJSON struct {
	Name           string
	PackageManager string
	Workspaces     []string
	Scripts        []script
}

var lifecycleScripts = map[string]bool{
	"dependencies": true, "install": true, "postinstall": true, "postpack": true,
	"postpublish": true, "postversion": true, "preinstall": true, "prepack": true,
	"prepare": true, "prepublish": true, "prepublishOnly": true, "preversion": true,
	"version": true,
}

func (Discovery) Scripts(root string, exclude []string) []config.Definition {
	root, _ = filepath.Abs(root)
	rootPackage, ok := readPackage(filepath.Join(root, "package.json"))
	if !ok {
		return nil
	}
	manager := packageManager(root, rootPackage)
	var out []config.Definition
	for _, pkg := range packages(root, rootPackage) {
		packageName := pkg.pkg.Name
		if packageName == "" {
			packageName = filepath.Base(pkg.dir)
		}
		relative, err := filepath.Rel(root, pkg.dir)
		if err != nil {
			continue
		}
		prefix := ""
		if relative != "." {
			prefix = packageName
		}
		for _, item := range filteredScripts(pkg.pkg.Scripts) {
			id := item.Name
			if prefix != "" {
				id = prefix + ":" + item.Name
			}
			if excluded(id, exclude) {
				continue
			}
			out = append(out, config.Definition{
				ID:          id,
				Name:        id,
				Description: item.Body,
				Command:     scriptCommand(manager, item.Name),
				Directory:   relative,
				Placement:   config.PlacementWindow,
				ExitMode:    config.ExitKeepOnError,
				Enabled:     true,
				Executable:  manager,
			})
		}
	}
	return out
}

type packageEntry struct {
	dir string
	pkg packageJSON
}

func packages(root string, rootPackage packageJSON) []packageEntry {
	entries := []packageEntry{{dir: root, pkg: rootPackage}}
	seen := map[string]bool{root: true}
	for _, pattern := range workspacePatterns(root, rootPackage) {
		if strings.HasPrefix(pattern, "!") {
			continue
		}
		matches, _ := filepath.Glob(filepath.Join(root, pattern, "package.json"))
		sort.Strings(matches)
		for _, path := range matches {
			abs, err := filepath.Abs(path)
			if err != nil || !strings.HasPrefix(abs, root+string(os.PathSeparator)) {
				continue
			}
			dir := filepath.Dir(abs)
			if seen[dir] {
				continue
			}
			pkg, ok := readPackage(abs)
			if ok {
				entries = append(entries, packageEntry{dir: dir, pkg: pkg})
				seen[dir] = true
			}
		}
	}
	return entries
}

func workspacePatterns(root string, rootPackage packageJSON) []string {
	patterns := append([]string{}, rootPackage.Workspaces...)
	path := filepath.Join(root, "pnpm-workspace.yaml")
	data, err := os.ReadFile(path)
	if err == nil {
		var doc struct {
			Packages []string `yaml:"packages"`
		}
		if yaml.Unmarshal(data, &doc) == nil {
			patterns = append(patterns, doc.Packages...)
		}
	}
	seen := map[string]bool{}
	out := make([]string, 0, len(patterns))
	for _, pattern := range patterns {
		pattern = strings.TrimSpace(pattern)
		if pattern != "" && !seen[pattern] {
			out = append(out, pattern)
			seen[pattern] = true
		}
	}
	return out
}

func filteredScripts(scripts []script) []script {
	names := make(map[string]bool, len(scripts))
	for _, item := range scripts {
		names[item.Name] = true
	}
	out := make([]script, 0, len(scripts))
	for _, item := range scripts {
		if lifecycleScripts[item.Name] || hookForExistingScript(item.Name, names) {
			continue
		}
		out = append(out, item)
	}
	return out
}

func hookForExistingScript(name string, names map[string]bool) bool {
	for _, prefix := range []string{"pre", "post"} {
		if strings.HasPrefix(name, prefix) && names[strings.TrimPrefix(name, prefix)] {
			return true
		}
	}
	return false
}

func packageManager(root string, pkg packageJSON) string {
	declared := strings.SplitN(pkg.PackageManager, "@", 2)[0]
	switch declared {
	case "pnpm", "yarn", "bun", "npm":
		return declared
	}
	if exists(filepath.Join(root, "pnpm-lock.yaml")) {
		return "pnpm"
	}
	if exists(filepath.Join(root, "yarn.lock")) {
		return "yarn"
	}
	if exists(filepath.Join(root, "bun.lock")) || exists(filepath.Join(root, "bun.lockb")) {
		return "bun"
	}
	return "npm"
}

func scriptCommand(manager, name string) string {
	escaped := shellwords.Escape(name)
	if manager == "npm" {
		return "npm run " + escaped
	}
	return manager + " " + escaped
}

func readPackage(path string) (packageJSON, bool) {
	data, err := os.ReadFile(path)
	if err != nil {
		return packageJSON{}, false
	}
	var raw map[string]json.RawMessage
	if err := json.Unmarshal(data, &raw); err != nil {
		return packageJSON{}, false
	}
	var pkg packageJSON
	_ = json.Unmarshal(raw["name"], &pkg.Name)
	_ = json.Unmarshal(raw["packageManager"], &pkg.PackageManager)
	pkg.Workspaces = parseWorkspaces(raw["workspaces"])
	pkg.Scripts = parseScripts(raw["scripts"])
	return pkg, true
}

func parseWorkspaces(raw json.RawMessage) []string {
	var arr []string
	if json.Unmarshal(raw, &arr) == nil {
		return arr
	}
	var obj struct {
		Packages []string `json:"packages"`
	}
	if json.Unmarshal(raw, &obj) == nil {
		return obj.Packages
	}
	return nil
}

func parseScripts(raw json.RawMessage) []script {
	if len(raw) == 0 || bytes.Equal(raw, []byte("null")) {
		return nil
	}
	decoder := json.NewDecoder(bytes.NewReader(raw))
	token, err := decoder.Token()
	if err != nil || token != json.Delim('{') {
		return nil
	}
	var scripts []script
	for decoder.More() {
		keyToken, err := decoder.Token()
		if err != nil {
			return nil
		}
		key, _ := keyToken.(string)
		var value any
		if err := decoder.Decode(&value); err != nil {
			return nil
		}
		scripts = append(scripts, script{Name: key, Body: toString(value)})
	}
	return scripts
}

func toString(value any) string {
	switch v := value.(type) {
	case string:
		return v
	case nil:
		return ""
	default:
		data, _ := json.Marshal(v)
		return string(data)
	}
}

func excluded(id string, patterns []string) bool {
	for _, pattern := range patterns {
		if fnmatch(pattern, id) {
			return true
		}
	}
	return false
}

func fnmatch(pattern, value string) bool {
	for _, expanded := range expandBraces(pattern) {
		re := globRegexp(expanded)
		if re.MatchString(value) {
			return true
		}
	}
	return false
}

func expandBraces(pattern string) []string {
	start := strings.Index(pattern, "{")
	if start < 0 {
		return []string{pattern}
	}
	end := strings.Index(pattern[start:], "}")
	if end < 0 {
		return []string{pattern}
	}
	end += start
	prefix, suffix := pattern[:start], pattern[end+1:]
	var out []string
	for _, choice := range strings.Split(pattern[start+1:end], ",") {
		for _, rest := range expandBraces(suffix) {
			out = append(out, prefix+choice+rest)
		}
	}
	return out
}

func globRegexp(pattern string) *regexp.Regexp {
	var b strings.Builder
	b.WriteString("^")
	inClass := false
	for _, r := range pattern {
		switch {
		case inClass:
			b.WriteRune(r)
			if r == ']' {
				inClass = false
			}
		case r == '*':
			b.WriteString(".*")
		case r == '?':
			b.WriteString(".")
		case r == '[':
			inClass = true
			b.WriteRune(r)
		default:
			b.WriteString(regexp.QuoteMeta(string(r)))
		}
	}
	b.WriteString("$")
	return regexp.MustCompile(b.String())
}

func exists(path string) bool {
	_, err := os.Stat(path)
	return err == nil
}
